pkg/sql/plan2/build_function.go

```diff
@@ -0,0 +1,263 @@
+	"github.com/matrixorigin/matrixone/pkg/sql/parsers/tree"
+)
+
+func checkFloatType(typ plan.Type_TypeId, alias string) error {
+	switch typ {
+	case plan.Type_FLOAT32:
+		fallthrough
+	case plan.Type_FLOAT64:
+		return nil
+	default:
+		return errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("type error: arg '%v' is not float", alias))
+	}
+}
+
+func checkIntType(typ plan.Type_TypeId, alias string) error {
+	switch typ {
+	case plan.Type_INT8:
+		fallthrough
+	case plan.Type_INT16:
+		fallthrough
+	case plan.Type_INT32:
+		fallthrough
+	case plan.Type_INT64:
+		fallthrough
+	case plan.Type_INT128:
+		fallthrough
+	case plan.Type_UINT8:
+		fallthrough
+	case plan.Type_UINT16:
+		fallthrough
+	case plan.Type_UINT32:
+		fallthrough
+	case plan.Type_UINT64:
+		fallthrough
+	case plan.Type_UINT128:
+		return nil
+	default:
+		return errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("type error: arg '%v' is not int", alias))
+	}
+}
+
+func checkDecimalType(typ plan.Type_TypeId, alias string) error {
+	switch typ {
+	case plan.Type_DECIMAL:
+		fallthrough
+	case plan.Type_DECIMAL64:
+		fallthrough
+	case plan.Type_DECIMAL128:
+		return nil
+	default:
+		return errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("type error: arg '%v' is not decimal", alias))
+	}
+}
+
+func checkNumberType(typ plan.Type_TypeId, alias string) error {
+	err := checkIntType(typ, alias)
+	if err == nil {
+		return nil
+	}
+	err = checkFloatType(typ, alias)
+	if err == nil {
+		return nil
+	}
+	err = checkDecimalType(typ, alias)
+	if err != nil {
+		return errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("type error: arg '%v' is not number", alias))
+	}
+	return nil
+}
+
+func checkTimeType(typ plan.Type_TypeId, alias string) error {
+	switch typ {
+	case plan.Type_DATE:
+		fallthrough
+	case plan.Type_TIME:
+		fallthrough
+	case plan.Type_DATETIME:
+		fallthrough
+	case plan.Type_TIMESTAMP:
+		fallthrough
+	case plan.Type_INTERVAL:
+		return nil
+	default:
+		return errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("type error: arg '%v' is not time", alias))
+	}
+}
+
+func checkFunctionArgs(fun *FunctionSig, args []*plan.Expr) error {
+	for idx, typ := range fun.ArgType {
+		argType := fun.ArgTypeClass[typ]
+		switch argType {
+		case plan.Type_ANY:
+			//do nothing
+		case plan.Type_ANYFLOAT:
+			return checkFloatType(args[idx].Typ.Id, args[idx].Alias)
+		case plan.Type_ANYINT:
+			return checkIntType(args[idx].Typ.Id, args[idx].Alias)
+		case plan.Type_ANYNUMBER:
+			return checkNumberType(args[idx].Typ.Id, args[idx].Alias)
+		case plan.Type_ANYTIME:
+			return checkTimeType(args[idx].Typ.Id, args[idx].Alias)
+		default:
+			if argType != args[idx].Typ.Id {
+				return errors.New(errno.SyntaxErrororAccessRuleViolation, fmt.Sprintf("type error: arg '%v' is not %v", args[idx].Alias, plan.Type_TypeId_name[int32(argType)]))
+			}
+		}
+	}
+	return nil
+}
+
+func coverArgToHigh(args []*plan.Expr) error {
```

Comment:
I think you need to add a cast function -- basically, 
int32 + int64      ->    translate into    int64(int32) + int64

Reply:
It is probably better to have a implicit cast table, saying type x can be cast to type y implicitly.   Then we look at this table.

For example, we know int8 -> int16 is ok, uint8 -> uint16 is ok, uint8 -> int16 is also ok. etc etc.

Reply:
type check and type cast will be done in next commit/PR，I need more research and design work。
And I got the  suggestion